sdk/storage/inc/common/concurrent_transfer.hpp

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <atomic>
+#include <cstdlib>
+#include <functional>
+#include <future>
+
+namespace Azure { namespace Storage { namespace Details {
+
+  inline void ConcurrentTransfer(
+      int64_t offset,
+      int64_t length,
+      int64_t chunkSize,
+      int concurrency,
+      // offset, length, chunk id, number of chunks
+      std::function<void(int64_t, int64_t, int64_t, int64_t)> transferFunc)
+  {
+    std::atomic<int> numWorkingThreads{concurrency};
+    std::atomic<int> nextChunkId{0};
+    std::atomic<bool> failed{false};
+
+    const auto numChunks = (length + chunkSize - 1) / chunkSize;
+
+    auto threadFunc = [&]() {
+      while (true)
+      {
+        int chunkId = nextChunkId.fetch_add(1);
+        if (chunkId >= numChunks || failed)
+        {
+          break;
+        }
+        int64_t chunkOffset = offset + chunkSize * chunkId;
+        int64_t chunkLength = std::min(length - chunkSize * chunkId, chunkSize);
+        try
+        {
+          transferFunc(chunkOffset, chunkLength, chunkId, numChunks);
+        }
+        catch (std::exception&)
+        {
+          if (failed.exchange(true) == false)
+          {
+            numWorkingThreads.fetch_sub(1);
+            throw;
+          }
+        }
+      }
+      numWorkingThreads.fetch_sub(1);
+    };
+
+    std::vector<std::future<void>> threadHandles;
+    for (int i = 0; i < concurrency - 1; ++i)
+    {
+      threadHandles.emplace_back(std::async(std::launch::async, threadFunc));
+    }
+    threadFunc();
+    for (auto& handle : threadHandles)
+    {
+      handle.get();
+    }
+
+    if (numWorkingThreads != 0)
+    {
+      std::abort();
```

Comment:
I believe even in a thread leak situation like this, we still should not call std::abort(). Please throw an exception instead.

Reply:
It's expressly stated in our guidelines that, in this situation, we should crash.

https://azure.github.io/azure-sdk/cpp_design.html#post-conditions

Reply:
This is not a post-condition according to the definition of post-condition:
> Post-Conditions
Post-Condition violations happen when some function didn’t do the correct thing, these are always bugs in the function itself, and users shouldn’t be expected to handle them.

I am not sure when this will happen. I would not even examine this value, to be honest.

Reply:
This is never supposed to happen. If it did happen, it means this function didn't do the correct thing, there was a bug in it. So by definition it's a post-condition check.

Reply:
> I would not even examine this value, to be honest.

Yes, it's totally reasonable if you don't want to check this value.

But that doesn't mean post-condition check is unnecessary. Since our guidelines mention how we should do post-condition check, I think that implies that post-condition check is not prohibited. So I suppose it's okay to do it here.